source/extensions/filters/network/dubbo_proxy/serialization.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+
+#include "common/singleton/const_singleton.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace DubboProxy {
+
+/**
+ * Names of available serialization implementations.
+ */
+class SerializerNameValues {
+public:
+  // hessian serialization
+  const std::string HESSIAN = "hessian";
+  // json serialization
+  const std::string JSON = "json";
+};
+
+typedef ConstSingleton<SerializerNameValues> SerializerNames;
+
+/**
+ * RpcInvocation represent an rpc call
+ * See
+ * https://github.com/apache/incubator-dubbo/blob/master/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcInvocation.java
+ */
+class RpcInvocation {
+public:
+  virtual ~RpcInvocation() {}
+  virtual const std::string& getMethodName() const PURE;
+  virtual const std::string& getServiceName() const PURE;
+  virtual const std::string& getServiceVersion() const PURE;
+};
+
+typedef std::unique_ptr<RpcInvocation> RpcInvocationPtr;
+
+/**
+ * RpcResult represent the result of an rpc call
+ * See
+ * https://github.com/apache/incubator-dubbo/blob/master/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/RpcResult.java
+ */
+class RpcResult {
+public:
+  virtual ~RpcResult() {}
+  virtual bool hasException() const PURE;
+};
+
+typedef std::unique_ptr<RpcResult> RpcResultPtr;
+
+class SerializationCallbacks {
+public:
+  virtual ~SerializationCallbacks() {}
+  virtual void onRpcInvocation(RpcInvocationPtr&& invo) PURE;
+  virtual void onRpcResult(RpcResultPtr&& res) PURE;
+};
+
+class Serialization {
+public:
+  virtual ~Serialization() {}
+  /**
+   * Return this serialization's name
+   *
+   * @return std::string containing the serialization name.
+   */
+  virtual const std::string& name() const PURE;
+  /**
+   * deserialize an rpc call, potentially invoking callbacks
+   * If successful, the RpcInvocation removed from the buffer
+   *
+   * @param buffer the currently buffered dubbo data
+   * @body_size the complete RpcInvocation size
+   * @throws EnvoyException if the data is not valid for this serialization
+   */
+  virtual void deserializeRpcInvocation(Buffer::Instance& buffer, size_t body_size) PURE;
```

Comment:
I don't know why you need a callback. Can this just return `RpcInvocationPtr`? Looked at #4200 too but I don't think there is a gain from this structure.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.

Reply:
The purpose of using callback here is to decouple the decoder and filter. All stats statistics are implemented in the filter. The decoder is only responsible for protocol parsing, and then the callback will parse the parsed result to the filter.